pkg/agent/controller/l7visibility/l7visibility_controller_test.go

```diff
@@ -0,0 +1,546 @@
+				mockOFClient.EXPECT().UninstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod1NN))
+			},
+		}, {
+			name:                               "Remove flows for deletedPod",
+			Pod:                                newPodObject("test-pod2", "test-ns1", annotationsCorrectIngress),
+			deletePod:                          true,
+			expectedL7PodNNDirAfterFlowRemoved: map[string]v1alpha2.Direction{pod1NN: directionIngress},
+			expectedUninstallCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().UninstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod2NN))
+			},
+		},
+	}
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newFakeControllerAndWatcher(t, []runtime.Object{pod1, pod2, testNS1}, interfaces)
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+			c.startInformers(stopCh)
+			assert.Eventuallyf(t, func() bool {
+				ns, _ := c.namespaceLister.List(labels.Everything())
+				return len(c.localPodInformer.GetIndexer().List()) == 2 && len(ns) == 1
+			}, 1*time.Second, 10*time.Millisecond, "Pods should be added to Informers")
+			expectedInstallCalls(c.mockOFClient)
+			waitEvents(t, 2, c)
+			for i := 0; i < 2; i++ {
+				item, _ := c.queue.Get()
+				err := c.syncPod(item.(string))
+				require.NoError(t, err)
+				c.queue.Done(item)
+			}
+			if tt.deletePod {
+				//Delete Pod
+				err := c.client.CoreV1().Pods(tt.Pod.Namespace).Delete(context.TODO(), tt.Pod.Name, metav1.DeleteOptions{})
+				require.NoError(t, err)
+			} else {
+				// Update Pods with no annotations
+				_, err := c.client.CoreV1().Pods(tt.Pod.Namespace).Update(context.TODO(), tt.Pod, metav1.UpdateOptions{})
+				require.NoError(t, err)
+			}
+
+			tt.expectedUninstallCalls(c.mockOFClient)
+
+			waitEvents(t, 1, c)
+			item, _ := c.queue.Get()
+			err := c.syncPod(item.(string))
+			require.NoError(t, err)
+			require.Equal(t, tt.expectedL7PodNNDirAfterFlowRemoved, c.podToDirectionMap)
+			c.queue.Done(item)
+		})
+	}
+}
+
+func TestNamespaceUpdate(t *testing.T) {
+	var targetPort uint32
+	testNS1 := newNamespaceObject("test-ns1", annotationsEmpty)
+	testNS2 := newNamespaceObject("test-ns2", annotationsEmpty)
+	pod1 := newPodObject("test-pod1", "test-ns1", annotationsEmpty)
+	pod2 := newPodObject("test-pod2", "test-ns1", annotationsEmpty)
+	pod3 := newPodObject("test-pod3", "test-ns2", annotationsEmpty)
+	pod4 := newPodObject("test-pod4", "test-ns2", annotationsCorrectIngress)
+	pod1NN := k8s.NamespacedName(pod1.Namespace, pod1.Name)
+	pod2NN := k8s.NamespacedName(pod2.Namespace, pod2.Name)
+	pod3NN := k8s.NamespacedName(pod3.Namespace, pod3.Name)
+	podInterface1 := newPodInterface(pod1.Name, pod1.Namespace, int32(pod1OFPort))
+	podInterface2 := newPodInterface(pod2.Name, pod2.Namespace, int32(pod2OFPort))
+	podInterface3 := newPodInterface(pod3.Name, pod3.Namespace, int32(pod3OFPort))
+	podInterface4 := newPodInterface(pod4.Name, pod4.Namespace, int32(pod4OFPort))
```

Comment:
by this point, I feel like we have a lot of duplicate declarations across all the tests?

Reply:
interfaces and namespacedName have been moved to vars, imo we should keep the pod and namespace objects in the function scope.